vault/core.go

```diff
@@ -909,6 +917,81 @@ func (c *Core) unsealInternal(masterKey []byte) (bool, error) {
 		c.logger.Info("core: vault is unsealed")
 	}
 
+	type versionStruct struct {
+		Version string `json:"version"`
+	}
+
+	// check version to make sure a downgrade hasn't happened
+	ver, err := c.barrier.Get(coreDataVersionPath)
+	if err != nil {
+		c.logger.Error("core: data version check failed", "error", err)
+		c.barrier.Seal()
+		c.logger.Warn("core: vault is sealed")
+		return false, err
+	}
+
+	persistVersion := false
+	verStruct := &versionStruct{}
+
+	localVer, err := gov.NewVersion(version.GetVersion().VersionNumber())
+	if err != nil {
+		// Don't do anything; this could be an unknown case such as during
+		// tests; we also don't want to persist the local version if we can't
+		// parse it
+		c.logger.Warn("core: could not perform data version check", "local_version", version.GetVersion().VersionNumber())
+		goto VERSION_DONE
+	}
+
+	if ver == nil {
+		persistVersion = true
+	} else {
+		err = jsonutil.DecodeJSON(ver.Value, verStruct)
+		if err != nil {
+			c.logger.Warn("core: could not decode version information", "version_val", string(ver.Value))
+			goto VERSION_DONE
+		}
+		dataVer, err := gov.NewVersion(verStruct.Version)
+		if err != nil {
+			c.logger.Warn("core: could not perform data version check", "saved_version", verStruct.Version)
+			goto VERSION_DONE
+		}
+		if localVer.LessThan(dataVer) {
```

Comment:
Nevermind, I was reading through the code once more and got what it's trying to do. We only need to handle the case where they are not equal, which we would either reject if lower, or persist if higher.